src/components/fast_clock.c

```diff
@@ -0,0 +1,253 @@
+// The clock is defined on div8. That means that the clock is triggered
+// every 50ms.
+#define FAST_CLOCK_TIMER_DELAY 50000
+Tc *fast_clock_timer;
+
+// ----------------------------------------------------------------------------
+//
+void fast_clock_init_timer(Tc *timer, uint32_t pm_tmr_mask, uint32_t gclock_tmr_id, uint32_t nvic_irqn)
+{
+  fast_clock_timer = timer;
+
+  // Enable clock for fast clock, without prescaler
+  PM->APBCMASK.reg |= pm_tmr_mask;
+  GCLK->CLKCTRL.reg =
+    GCLK_CLKCTRL_ID(gclock_tmr_id)
+    | GCLK_CLKCTRL_CLKEN
+    | GCLK_CLKCTRL_GEN(0);
+
+  /* CTRLA register:
+   *   PRESCSYNC: 0x02  RESYNC
+   *   RUNSTDBY:        Ignored
+   *   PRESCALER: 0x03  DIV8, each tick is be 1 us
+   *   WAVEGEN:   0x01  MFRQ, zero counter on match
+   *   MODE:      0x00  16 bits timer
+   */
+  fast_clock_timer->COUNT16.CTRLA.reg =
+    TC_CTRLA_PRESCSYNC_RESYNC
+    | TC_CTRLA_PRESCALER_DIV8
+    | TC_CTRLA_WAVEGEN_MFRQ
+    | TC_CTRLA_MODE_COUNT16;
+
+  fast_clock_timer->COUNT16.INTENSET.reg = TC_INTENSET_MC(1);
+  NVIC_EnableIRQ(nvic_irqn);
+
+  fast_clock_timer->COUNT16.COUNT.reg = 0;
+  fast_clock_timer->COUNT16.CC[0].reg = FAST_CLOCK_TIMER_DELAY;
+  fast_clock_timer->COUNT16.CTRLA.reg |= TC_CTRLA_ENABLE;
+}
+
+// ----------------------------------------------------------------------------
+// sets the values for being a master: whether it is a
+void fast_clock_set_as_master(uint8_t id1, uint8_t id2, uint16_t intermessage_delay)
+{
+  fast_clock_status.master = true;
+  fast_clock_status.id1 = id1;
+  fast_clock_status.id2 = id2;
+  fast_clock_status.intermessage_delay = intermessage_delay;
+}
+
+// ----------------------------------------------------------------------------
+// sets the thing as a slave.
+void fast_clock_set_as_slave(void)
+{
+  fast_clock_status.master = false;
+}
+
+//----------------------------------------------------------------------------
+void fast_clock_set_time(FAST_CLOCK_TIME_Type time)
+{
+  // Set the time
+  current_time = time;
+  // Reset the millisecond counter
+  fast_clock_status.millisecond = 0;
+  // Notify the update!
+  fast_clock_handle_update(current_time);
+}
+
+//----------------------------------------------------------------------------
+// this function sets the clock rate.
+void fast_clock_set_rate(uint8_t rate)
+{
+  fast_clock_status.rate = rate;
+}
+
+//----------------------------------------------------------------------------
+// this function processes the clock information sent on the loconet
+void loconet_rx_fast_clock(uint8_t *data, uint8_t length)
+{
+  if (length < 8)
+  {
+    return;
+  }
+
+  if (!(data[7] == 1)) {
+    // the message is not a correct clock tick!
+    return;
+  }
+
+  //1st byte of data is the clock rate
+  fast_clock_set_rate(data[0]);
+
+  current_time.minute = data[3] - (128-60);
+  current_time.hour = data[5] >= (128-24) ? data[5] - (128-24) : data[5] % 24;
+  current_time.day = data[6] % 7;
+  current_time.second = 0;
```

Comment:
Above you have the flow `minute`, `hour`, `day` and then `second`. To keep the flow, you can place `second` before `minute`. Same for the `millisecond` below.